mm2src/coins/tx_history_storage/wasm/tx_history_storage_v2.rs

```diff
@@ -0,0 +1,484 @@
+use crate::my_tx_history_v2::{GetHistoryResult, RemoveTxResult, TxHistoryStorage};
+use crate::tx_history_storage::wasm::tx_history_db::{TxHistoryDb, TxHistoryDbLocked};
+use crate::tx_history_storage::wasm::{WasmTxHistoryError, WasmTxHistoryResult};
+use crate::tx_history_storage::{token_id_from_tx_type, ConfirmationStatus, CreateTxHistoryStorageError,
+                                FilteringAddresses, GetTxHistoryFilters, WalletId};
+use crate::{CoinsContext, TransactionDetails};
+use async_trait::async_trait;
+use common::PagingOptionsEnum;
+use itertools::Itertools;
+use mm2_core::mm_ctx::MmArc;
+use mm2_db::indexed_db::{BeBigUint, DbUpgrader, MultiIndex, OnUpgradeResult, SharedDb, TableSignature};
+use mm2_err_handle::prelude::*;
+use rpc::v1::types::Bytes as BytesJson;
+use serde_json::{self as json, Value as Json};
+
+impl WalletId {
+    /// If [`WalletId::hd_wallet_rmd160`] is not specified,
+    /// we need to exclude transactions of each HD wallet by specifying an empty `hd_wallet_rmd160`.
+    fn hd_wallet_rmd160_or_exclude(&self) -> String {
+        self.hd_wallet_rmd160.map(|hash| hash.to_string()).unwrap_or_default()
+    }
+}
+
+#[derive(Clone)]
+pub struct IndexedDbTxHistoryStorage {
+    db: SharedDb<TxHistoryDb>,
+}
+
+impl IndexedDbTxHistoryStorage {
+    pub fn new(ctx: &MmArc) -> MmResult<Self, CreateTxHistoryStorageError>
+    where
+        Self: Sized,
+    {
+        let coins_ctx = CoinsContext::from_ctx(ctx).map_to_mm(CreateTxHistoryStorageError::Internal)?;
+        Ok(IndexedDbTxHistoryStorage {
+            db: coins_ctx.tx_history_db.clone(),
+        })
+    }
+}
+
+#[async_trait]
+impl TxHistoryStorage for IndexedDbTxHistoryStorage {
+    type Error = WasmTxHistoryError;
+
+    async fn init(&self, _wallet_id: &WalletId) -> MmResult<(), Self::Error> { Ok(()) }
+
+    async fn is_initialized_for(&self, _wallet_id: &WalletId) -> MmResult<bool, Self::Error> { Ok(true) }
+
+    /// Adds multiple transactions to the selected coin's history
+    /// Also consider adding tx_hex to the cache during this operation
+    async fn add_transactions_to_history<I>(&self, wallet_id: &WalletId, transactions: I) -> MmResult<(), Self::Error>
+    where
+        I: IntoIterator<Item = TransactionDetails> + Send + 'static,
+        I::IntoIter: Send,
+    {
+        let locked_db = self.lock_db().await?;
+        let db_transaction = locked_db.get_inner().transaction().await?;
+        let history_table = db_transaction.table::<TxHistoryTableV2>().await?;
+        let cache_table = db_transaction.table::<TxCacheTableV2>().await?;
+
+        for tx in transactions {
+            let history_item = TxHistoryTableV2::from_tx_details(wallet_id.clone(), &tx)?;
+            history_table.add_item(&history_item).await?;
```

Comment:
If one of the tx insertion fails, will the entire DB transaction rollback, or will we have a partial insertion performed?

Reply:
As it's said in the [doc](https://developer.mozilla.org/en-US/docs/Web/API/IDBTransaction#transaction_failures), `IdbTransaction.abort` has to be fired on any invalid request associated with the transaction.

Reply:
Do you have to call it explicitly?

Reply:
Oh no, the `abort` event is fired immediately on an error, so there is no need to call it manually. I can add a test that will check if items are not in the storage if the transaction is aborted.

Reply:
Yes, this would be great :)

Reply:
I added the [indexed_db::test_transaction_abort_on_error](https://github.com/KomodoPlatform/atomicDEX-API/blob/hd-wallet-tx-history/mm2src/mm2_db/src/indexed_db/indexed_db.rs#L1416) test.